### PR TITLE
NET-152 enrollment keys for non admins

### PR DESCRIPTION
**Pull request by TobiaszCudnik**
## Describe your changes

NET-152 enrollment keys for non admins
- fixed json parsing in `PUT /api/users/networks/user1`
- added `/api/enrollment_keys` for non-admins
- fixed expired JWT results in 500
- unit test for enrollment keys for users
- added `/api/hosts` for non-admins
  - using `GetHostNetworks`, only full-access hosts shown

## Checklist before requesting a review
- [ ] My changes affect only 10 files or less.
- [ ] I have performed a self-review of my code and tested it.
- [ ] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [ ] My functions are <= 80 lines.
- [ ] I have had my code reviewed by a peer.
- [ ] My unit tests pass locally.
- [ ] Netmaker is awesome.

### CODE REVIEW

File: controllers/enrollmentkeys.go [hunk 1]
```diff
@@ -34,24 +34,36 @@ func enrollmentKeyHandlers(r *mux.Router) {
...
 //				200: getEnrollmentKeysSlice
 func getEnrollmentKeys(w http.ResponseWriter, r *http.Request) {
-	currentKeys, err := logic.GetAllEnrollmentKeys()
+	user, err := logic.GetUser(r.Header.Get("user"))
+	if err != nil {
+		logger.Log(0, r.Header.Get("user"), "failed to fetch user: ", err.Error())
+		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "internal"))
+		return
+	}
+	fmt.Println(user)
```
Comment (Aceix) [inline controllers/enrollmentkeys.go:43]: remove?

File: controllers/hosts.go [hunk 2]
```diff
@@ -52,12 +52,26 @@ func getHosts(w http.ResponseWriter, r *http.Request) {
...
+	ret := []models.ApiHost{}
 	apiHosts := logic.GetAllHostsAPI(currentHosts[:])
 	logger.Log(2, r.Header.Get("user"), "fetched all hosts")
-	logic.SortApiHosts(apiHosts[:])
+	for _, host := range apiHosts {
+		networks := logic.GetHostNetworks(host.ID)
+		if !logic.UserHasNetworksAccess(networks, user) {
+			continue
+		}
+		ret = append(ret, host)
```
Comment (Aceix) [inline controllers/hosts.go:70]: might be tricky. if hostA is connected to netA and netB, and user has access only to netA, then hostA info will not be loaded for him. i think it should be loaded in contrast to the enrollment key

**Review by Aceix — CHANGES_REQUESTED** [review]

File: controllers/enrollmentkeys.go (hunk 1 above)
Comment (TobiaszCudnik) [reply controllers/enrollmentkeys.go]: Fixed.

File: controllers/enrollmentkeys.go [hunk 3]
```diff
@@ -34,24 +34,36 @@ func enrollmentKeyHandlers(r *mux.Router) {
...
+	keys, err := logic.GetAllEnrollmentKeys()
+	// TODO drop double pointer
+	accessKeys := []*models.EnrollmentKey{}
```
[the remark this answers is not in the document]
Comment (TobiaszCudnik) [reply controllers/enrollmentkeys.go]: Fixed.

**Comment by theguy951357** [comment]
unable to make api calls to get the enrollment keys using the master key for authorization bearer. details in kitemaker ticket

File: controllers/hosts.go (hunk 2 above)
Comment (TobiaszCudnik) [reply controllers/hosts.go]: Ive implemented a partial per-host access and only the nodes for accessible networks are returned within the host object.

File: controllers/enrollmentkeys.go [hunk 4]
```diff
@@ -34,35 +34,45 @@ func enrollmentKeyHandlers(r *mux.Router) {
...
-	user, err := logic.GetUser(r.Header.Get("user"))
+	keys, err := logic.GetAllEnrollmentKeys()
 	if err != nil {
-		logger.Log(0, r.Header.Get("user"), "failed to fetch user: ", err.Error())
+		logger.Log(0, r.Header.Get("user"), "failed to fetch enrollment keys: ", err.Error())
 		logic.ReturnErrorResponse(w, r, logic.FormatError(err, "internal"))
 		return
 	}
-	keys, err := logic.GetAllEnrollmentKeys()
+	// handle masteradmin non-logged-in user
```
Comment (mattkasun) [inline controllers/enrollmentkeys.go:43]: if NetUserSecurityCheck was used as middleware vice SecurityCheck, this logic should not be required.

**Review by mattkasun — CHANGES_REQUESTED** [review]

**Review by theguy951357 — APPROVED** [review]

**Review by abhishek9686 — APPROVED** [review]

**Review by Aceix — APPROVED** [review]

**Pull request merged**